java/memory/src/main/java/org/apache/arrow/memory/LowCostIdentityHasMap.java

```diff
@@ -0,0 +1,338 @@
+ *  limitations under the License.
+ */
+
+package org.apache.arrow.memory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+/**
+ * Highly specialized IdentityHashMap that implements only partial
+ * Map APIs
+ * It incurs low initial cost (just two elements by default)
+ * It assumes Value includes the Key - Implements @ValueWithKeyIncluded iface
+ * that provides "getKey" method
+ * @param <V>
+ */
+public class LowCostIdentityHasMap<V extends ValueWithKeyIncluded> {
+
+  /*
+   * The internal data structure to hold values.
+   */
+  private Object[] elementData;
+
+  /* Actual number of values. */
+  private int size;
+
+  /*
+   * maximum number of elements that can be put in this map before having to
+   * rehash.
+   */
+  private int threshold;
+
+  private static final int DEFAULT_MIN_SIZE = 1;
+
+  /* Default load factor of 0.75; */
+  private static final int loadFactor = 7500;
+
+  /**
+   * Creates an Map with default expected maximum size.
+   */
+  public LowCostIdentityHasMap() {
+    this(DEFAULT_MIN_SIZE);
+  }
+
+  /**
+   * Creates an Map with the specified maximum size parameter.
+   *
+   * @param maxSize
+   *            The estimated maximum number of entries that will be put in
+   *            this map.
+   */
+  public LowCostIdentityHasMap(int maxSize) {
+    if (maxSize >= 0) {
+      this.size = 0;
+      threshold = getThreshold(maxSize);
+      elementData = newElementArray(computeElementArraySize());
+    } else {
+      throw new IllegalArgumentException();
+    }
+  }
+
+  private int getThreshold(int maxSize) {
+    // assign the threshold to maxSize initially, this will change to a
+    // higher value if rehashing occurs.
+    return maxSize > 2 ? maxSize : 2;
+  }
+
+  private int computeElementArraySize() {
+    int arraySize = (int) (((long) threshold * 10000) / loadFactor);
+    // ensure arraySize is positive, the above cast from long to int type
+    // leads to overflow and negative arraySize if threshold is too big
+    return arraySize < 0 ? -arraySize : arraySize;
+  }
+
+  /**
+   * Create a new element array
+   *
+   * @param s
+   *            the number of elements
+   * @return Reference to the element array
+   */
+  private Object[] newElementArray(int s) {
+    return new Object[s];
+  }
+
+  private V massageValue(Object value) {
+    return (V) ((value == null) ? null : value);
+  }
+
+  /**
+   * Removes all elements from this map, leaving it empty.
+   *
+   * @see #isEmpty()
+   * @see #size()
+   */
+  public void clear() {
+    size = 0;
+    for (int i = 0; i < elementData.length; i++) {
+      elementData[i] = null;
+    }
+  }
+
+  /**
+   * Returns whether this map contains the specified key.
+   *
+   * @param key
+   *            the key to search for.
+   * @return {@code true} if this map contains the specified key,
+   *         {@code false} otherwise.
+   */
+  public boolean containsKey(Object key) {
```

Comment:
Isn't there a way to make this typed without requiring two fields in the class type definition?

Reply:
The only thing that can be done is : public <K> V get(K key)
But in this case it is quite detached from "K" defined in 
ValueWithKeyIncluded<K>

Reply:
Hmmm - it does not like angle brackets in comments:
<code> 
public \<K\> V get(K key)
<code>